### PR TITLE
Adding resilience_mode field support to Cloud Composer

### PR DESCRIPTION
**Pull request by Khaledmohamedrefaat**
Add Cloud Composer Support for high resilience field 

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
composer: added `resilience_mode` field in `google_composer_environment`
```

Fixes [hashicorp/terraform-provider-google/issues/14751](https://github.com/hashicorp/terraform-provider-google/issues/14751)

### CODE REVIEW

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 1]
```diff
@@ -526,6 +526,39 @@ func TestAccComposerEnvironment_ComposerV2(t *testing.T) {
 	})
 }
 
+func TestAccComposerEnvironment_ComposerV2HighResilience(t *testing.T) {
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:529]: I feel like we don't need a separate test for just one new field, I suggest incorporating it into different test (for example into TestAccComposerEnvironment_ComposerV2.

**Review by kubasieron — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/resources/resource_composer_environment.go.erb [hunk 2]
```diff
@@ -802,14 +803,23 @@ func ResourceComposerEnvironment() *schema.Resource {
...
 						"environment_size": {
-              Type:         schema.TypeString,
-              Optional:     true,
-							Computed:			true,
-              ForceNew:     false,
-              AtLeastOneOf: composerConfigKeys,
-              ValidateFunc: validation.StringInSlice([]string{"ENVIRONMENT_SIZE_SMALL", "ENVIRONMENT_SIZE_MEDIUM", "ENVIRONMENT_SIZE_LARGE"}, false),
-              Description:  `The size of the Cloud Composer environment. This field is supported for Cloud Composer environments in versions composer-2.*.*-airflow-*.*.* and newer.`,
-            },
+							Type:         schema.TypeString,
```
[the remark this answers is not in the document]
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/resources/resource_composer_environment.go.erb:806]: Fixed Indentations (Also, Environment Size had wrong indentations so they were adjusted).

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 3]
```diff
@@ -1676,6 +1709,85 @@ resource "google_compute_subnetwork" "test" {
...
+        subnetwork       = google_compute_subnetwork.test.self_link
+        ip_allocation_policy {
+          cluster_ipv4_cidr_block = "10.0.0.0/16"
+        }
+      }
+
+      software_config {
+        image_version = "composer-2-airflow-2"
+<% unless version == "ga" -%>
+        cloud_data_lineage_integration {
```
[the remark this answers is not in the document]
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb]: Removed High Resilience test, and added the field to TestAccComposerEnvironment_ComposerV2.

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 4]
```diff
@@ -1676,6 +1709,85 @@ resource "google_compute_subnetwork" "test" {
 `, envName, network, subnetwork)
 }
 
+func testAccComposerEnvironment_composerV2HighResilience(envName, network, subnetwork string) string {
+	return fmt.Sprintf(`
+resource "google_composer_environment" "test" {
+  name   = "%s"
+  region = "us-east1"
```
[the remark this answers is not in the document]
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb]: Removed High Resilience test, and added the field to TestAccComposerEnvironment_ComposerV2.

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 5]
```diff
@@ -1676,6 +1709,85 @@ resource "google_compute_subnetwork" "test" {
...
+        }
+				<% unless version == "ga" -%>
+				triggerer {
+					cpu         = 0.5
+					memory_gb   = 2.0
+					count   		= 1
+				}
+				<% end -%>
+      }
+      environment_size = "ENVIRONMENT_SIZE_MEDIUM"
```
[the remark this answers is not in the document]
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb]: Removed High Resilience test, and added the field to TestAccComposerEnvironment_ComposerV2.

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 6]
```diff
@@ -1676,6 +1709,85 @@ resource "google_compute_subnetwork" "test" {
...
+      software_config {
+        image_version = "composer-2-airflow-2"
+<% unless version == "ga" -%>
+        cloud_data_lineage_integration {
+          enabled = true
+        }
+<% end -%>
+      }
+
+      workloads_config {
```
[the remark this answers is not in the document]
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb]: Removed High Resilience test, and added the field to TestAccComposerEnvironment_ComposerV2.

**Review by kubasieron — APPROVED** [review]

**Review by kubasieron — APPROVED** [review]

**Review by kubasieron — APPROVED** [review]

**Comment by roaks3** [comment]
It looks like there's a build error:
```
sh -c "'/home/runner/work/magic-modules/magic-modules/provider/scripts/gofmtcheck.sh'"
==> Checking that code complies with gofmt requirements...
Error: ./google/resource_composer_environment_test.go:530:1: expected declaration, found '}'
Error: ./google/resource_composer_environment_test.go:23[8](https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/5192309052/jobs/9361294667#step:8:9)4:1: raw string literal not terminated
```

**Review by kubasieron — APPROVED** [review]

**Comment by roaks3** [comment]
@Khaledmohamedrefaat can you confirm that you are able to view the error messages from the most recent `modular-magician` comment? If not I will need to copy the errors as they come up.

In this case:
```
    vcr_utils.go:148: Step 2/4 error: Error running apply: exit status 1
        
        Error: googleapi: Error 400: Found 2 problems:
        	1) In high resilience mode, you have to set the number of schedulers to 2.
        	2) Provided triggerers count should be either 0 or 2., badRequest
        
          with google_composer_environment.test,
          on terraform_plugin_test.tf line 2, in resource "google_composer_environment" "test":
           2: resource "google_composer_environment" "test" {
```

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 7]
```diff
@@ -1676,6 +1710,77 @@ resource "google_compute_subnetwork" "test" {
...
+				triggerer {
+					cpu         = 0.5
+					memory_gb   = 2.0
+					count   		= 2
+				}
+				<% end -%>
+      }
+      environment_size = "ENVIRONMENT_SIZE_MEDIUM"
+      resilience_mode = "HIGH_RESILIENCE"
+	  private_environment_config {
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1758]: is private config needed for high resilience?

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 8]
```diff
@@ -1676,6 +1710,77 @@ resource "google_compute_subnetwork" "test" {
...
+        }
+        web_server {
+          cpu         = 1.75
+          memory_gb   = 3.0
+          storage_gb   = 4.4
+        }
+        worker {
+          cpu         = 0.5
+          memory_gb   = 2.0
+          storage_gb   = 3.4
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1744]: fix "=" indentation

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 9]
```diff
@@ -1676,6 +1710,77 @@ resource "google_compute_subnetwork" "test" {
...
+        scheduler {
+          cpu         = 1.25
+          memory_gb   = 2.5
+          storage_gb   = 5.4
+          count       = 2
+        }
+        web_server {
+          cpu         = 1.75
+          memory_gb   = 3.0
+          storage_gb   = 4.4
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1739]: fix "=" indentation

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 10]
```diff
@@ -1676,6 +1710,77 @@ resource "google_compute_subnetwork" "test" {
...
+
+      software_config {
+        image_version = "composer-2-airflow-2"
+      }
+
+      workloads_config {
+        scheduler {
+          cpu         = 1.25
+          memory_gb   = 2.5
+          storage_gb   = 5.4
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1733]: fix "=" indentation

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 11]
```diff
@@ -1645,7 +1679,7 @@ resource "google_composer_environment" "test" {
 				triggerer {
 					cpu         = 0.5
 					memory_gb   = 2.0
-					count   		= 1
+					count   		= 2
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1682]: I suppose it doesn't need to be changed now

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 12]
```diff
@@ -526,6 +526,40 @@ func TestAccComposerEnvironment_ComposerV2(t *testing.T) {
...
+			{
+				ResourceName:			"google_composer_environment.test",
+				ImportState:			 true,
+				ImportStateVerify: true,
+			},
+			// This is a terrible clean-up step in order to get destroy to succeed,
+			// due to dangling firewall rules left by the Composer Environment blocking network deletion.
+			// TODO(dzarmola): Remove this check if firewall rules bug gets fixed by Composer.
+			{
+				PlanOnly:					 true,
```
Comment (kubasieron) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:554]: fix indentations in the test

**Review by kubasieron — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb (hunk 7 above)
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1758]: yes, HR requires private config.

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb (hunk 11 above)
Comment (Khaledmohamedrefaat) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1682]: Agreed, rolling back this line.

**Review by kubasieron — APPROVED** [review]

File: mmv1/third_party/terraform/resources/resource_composer_environment.go.erb [hunk 13]
```diff
@@ -802,14 +803,23 @@ func ResourceComposerEnvironment() *schema.Resource {
...
+							Optional:     true,
+							Computed:     true,
+							ForceNew:     false,
+							AtLeastOneOf: composerConfigKeys,
+							ValidateFunc: validation.StringInSlice([]string{"ENVIRONMENT_SIZE_SMALL", "ENVIRONMENT_SIZE_MEDIUM", "ENVIRONMENT_SIZE_LARGE"}, false),
+							Description:  `The size of the Cloud Composer environment. This field is supported for Cloud Composer environments in versions composer-2.*.*-airflow-*.*.* and newer.`,
+						},
+						"resilience_mode": {
+							Type:         	schema.TypeString,
```
Comment (roaks3) [inline mmv1/third_party/terraform/resources/resource_composer_environment.go.erb:815]: Nit: inconsistent alignment of values in this block

File: mmv1/third_party/terraform/resources/resource_composer_environment.go.erb [hunk 14]
```diff
@@ -802,14 +803,23 @@ func ResourceComposerEnvironment() *schema.Resource {
...
+							Optional:		true,
+							Computed:		true,
+							ForceNew:		true,
+							AtLeastOneOf: composerConfigKeys,
+							ValidateFunc: validation.StringInSlice([]string{"RESILIENCE_MODE_UNSPECIFIED", "HIGH_RESILIENCE"}, false),
```
Comment (roaks3) [inline mmv1/third_party/terraform/resources/resource_composer_environment.go.erb:820]: We typically don't need `*_UNSPECIFIED` values in Terraform, because the user can just not include the field in the config. Do you know if it is necessary in this case?

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 15]
```diff
@@ -1676,6 +1710,76 @@ resource "google_compute_subnetwork" "test" {
...
+				storage_gb  = 4.4
+			}
+			worker {
+				cpu         = 0.5
+				memory_gb   = 2.0
+				storage_gb  = 3.4
+				min_count   = 2
+				max_count   = 5
+			}
+			<% unless version == "ga" -%>
```
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1748]: Nit: is this relevant to resilience_mode? If not, it might be simplest to just remove this block.

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb [hunk 16]
```diff
@@ -1676,6 +1710,76 @@ resource "google_compute_subnetwork" "test" {
...
+				memory_gb   = 2.0
+				count       = 2
+			}
+			<% end -%>
+		}
+		environment_size = "ENVIRONMENT_SIZE_MEDIUM"
+		resilience_mode = "HIGH_RESILIENCE"
+		private_environment_config {
+			enable_private_endpoint                  = true
+			cloud_composer_network_ipv4_cidr_block   = "10.3.192.0/24"
```
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1760]: I'm a little out of my depth so want to make sure: is there any risk of these cidr ranges conflicting with other resources in our various tests environments? (for example, if this test runs at the same time as the test above, testAccComposerEnvironment_composerV2)

File: mmv1/third_party/terraform/website/docs/r/composer_environment.html.markdown [hunk 17]
```diff
@@ -679,6 +679,12 @@ The `config` block supports:
   environment size are `ENVIRONMENT_SIZE_SMALL`, `ENVIRONMENT_SIZE_MEDIUM`,
   and `ENVIRONMENT_SIZE_LARGE`.
 
+* `resilience_mode` -
+  (Optional, Cloud Composer 2.1.15 or newer only)
+  The resilience mode states whether high resilience is enabled for 
+  the environment or not. Values for resilience mode are 
+  `RESILIENCE_MODE_UNSPECIFIED` and `HIGH_RESILIENCE`.
```
Comment (roaks3) [inline mmv1/third_party/terraform/website/docs/r/composer_environment.html.markdown:686]: If `RESILIENCE_MODE_UNSPECIFIED` is removed, we can remove it from the docs here too

**Review by roaks3 — COMMENTED** [review]
LGTM and tests pass. I had a few small questions/suggestions, but nothing major.

File: mmv1/third_party/terraform/resources/resource_composer_environment.go.erb (hunk 13 above)
Comment (PeWu) [reply mmv1/third_party/terraform/resources/resource_composer_environment.go.erb]: Done

File: mmv1/third_party/terraform/resources/resource_composer_environment.go.erb (hunk 14 above)
Comment (PeWu) [reply mmv1/third_party/terraform/resources/resource_composer_environment.go.erb]: I removed the `*_UNSPECIFIED` value

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb (hunk 15 above)
Comment (PeWu) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb]: Removed. This should make the test cleaner.

File: mmv1/third_party/terraform/website/docs/r/composer_environment.html.markdown (hunk 17 above)
Comment (PeWu) [reply mmv1/third_party/terraform/website/docs/r/composer_environment.html.markdown]: Done

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb (hunk 16 above)
Comment (PeWu) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1753]: I'm not sure if there is a risk. Nonetheless, these IP ranges have been duplicated several times already in these tests. I expect that duplicating them once more shouldn't change much.

**Comment by PeWu** [comment]
Thanks! I addressed the comments. PTAL.

File: mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb (hunk 16 above)
Comment (roaks3) [reply mmv1/third_party/terraform/tests/resource_composer_environment_test.go.erb:1753]: Oh alright, I hadn't noticed that, but agreed it shouldn't be an issue then.

**Comment by roaks3** [comment]
This is ready to merge, but we'll need to resolve the conflicts first, and then just double check that tests still pass.

**Comment by PeWu** [comment]
**VCR-test** is marked as failed but when I click _Details_, it shows as successful in Cloud Build. What can be the issue here?

**Comment by Khaledmohamedrefaat** [comment]
@roaks3 I've tried running `TestAccComposerEnvironment_ComposerV2HighResilience` locally without any errors so it seems this error was just a flake, Let's retrigger the checks again.

**Review by roaks3 — APPROVED** [review]

**Comment by roaks3** [comment]
Merging because the tests passed this time, and the missing role that caused the failure should not be an issue under normal circumstances.
